src/cache.cc

```diff
@@ -1464,6 +1485,119 @@ int CACHE::kpc_prefetch_line(uint64_t base_addr, uint64_t pf_addr, int pf_fill_l
     return 0;
 }
 
+int CACHE::va_prefetch_line(uint64_t ip, uint64_t pf_addr, int pf_fill_level, uint32_t prefetch_metadata)
+{
+  if(pf_addr == 0)
+    {
+      cout << "va_prefetch_line() pf_addr cannot be 0! exiting" << endl;
+      assert(0);
+    }
+
+  pf_requested++;
+  if(VAPQ.occupancy < VAPQ.SIZE)
+    {
+      // generate new prefetch request packet
+      PACKET pf_packet;
+      pf_packet.fill_level = pf_fill_level;
+      pf_packet.pf_origin_level = fill_level;
+      if(pf_fill_level == FILL_L1)
+	{
+	  pf_packet.fill_l1d = 1;
+	}
+      pf_packet.pf_metadata = prefetch_metadata;
+      pf_packet.cpu = cpu;
+      pf_packet.v_address = pf_addr >> LOG2_BLOCK_SIZE;
+      pf_packet.address = pf_addr >> LOG2_BLOCK_SIZE; // make address == v_address before translation just so we can use VAPQ's check_queue() function
+	pf_packet.full_v_addr = pf_addr;
+      pf_packet.full_addr = pf_addr;
+      pf_packet.ip = ip;
+      pf_packet.type = PREFETCH;
+      pf_packet.event_cycle = 0;
+
+      int vapq_index = VAPQ.check_queue(&pf_packet);
+      if(vapq_index != -1)
+	{
+	  // there's already a VA prefetch to this cache line
+	  return 1;
+	}
+
+      // add the packet to the virtual address space prefetching queue
+      int index = VAPQ.tail;
+      VAPQ.entry[index] = pf_packet;
+      VAPQ.occupancy++;
+      VAPQ.tail++;
+      if (VAPQ.tail >= VAPQ.SIZE)
+	{
+	  VAPQ.tail = 0;
+	}
+
+      return 1;
+    }
+
+  return 0;
+}
+
+void CACHE::va_translate_prefetches()
+{
+  // move translated prefetches from the VAPQ to the regular PQ
+  uint32_t vapq_index = VAPQ.head;
+  if (PQ.occupancy < PQ.SIZE)
+    {
+      for(uint32_t i=0; i<VAPQ.SIZE; i++)
+	{
+	  // identify a VA prefetch that is fully translated
+	  if((VAPQ.entry[vapq_index].address != 0) && (VAPQ.entry[vapq_index].address != VAPQ.entry[vapq_index].v_address))
+	    {
+	      // move the translated prefetch over to the regular PQ
+	      add_pq(&VAPQ.entry[vapq_index]);
+
+	      // remove the prefetch from the VAPQ
+	      VAPQ.remove_queue(&VAPQ.entry[vapq_index]);
+
+	      break;
+	    }
+	  vapq_index++;
+	  if(vapq_index >= VAPQ.SIZE)
+	    {
+	      vapq_index = 0;
+	    }
+	}
+    }
+
+  // TEMPORARY SOLUTION: mark prefetches as translated after a fixed latency
+  vapq_index = VAPQ.head;
+  for(uint32_t i=0; i<VAPQ.SIZE; i++)
+    {
+      if((VAPQ.entry[vapq_index].address == VAPQ.entry[vapq_index].v_address) && (VAPQ.entry[vapq_index].event_cycle <= current_core_cycle[cpu]))
```

Comment:
I think this could be a dangerous condition. There is an (unlikely) chance that the physical address and the virtual address could be identical. Could we instead include a `bool va_translated` flag to the `PACKET` class?